File: src/lib/providers/codex.ts
import { mkdtemp, readFile, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";

import type { ProviderAdapter, ProviderStatus } from "../types";
import {
  buildChapterChatPrompt,
  buildSummaryPrompt,
  SUMMARY_SCHEMA,
} from "../prompt";
import { runCommand } from "../process";
import { getEffectiveSettings } from "../settings";
import { parseSummaryJson, unavailableStatus } from "./shared";

const PROVIDER_NAME = "Codex";

export const codexProvider: ProviderAdapter = {
  id: "codex",
  name: PROVIDER_NAME,
  kind: "cli",

  async getStatus(): Promise<ProviderStatus> {
    const settings = getEffectiveSettings();
    const command = settings.cli.codexCommand || "codex";

    try {
      const result = await runCommand(command, ["login", "status"], {
        timeoutMs: 10_000,
      });
      const text = `${result.stdout}\n${result.stderr}`;
      const loggedIn = /logged in/i.test(text);

      return {
        id: "codex",
        name: PROVIDER_NAME,
        kind: "cli",
        available: true,
        loggedIn,
        detail: loggedIn
          ? normalizeStatusDetail(text)
          : `未ログイン。\`${command} login\` が必要です。`,
      };
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code === "ENOENT") {
        return unavailableStatus(
          "codex",
          PROVIDER_NAME,
          "cli",
          `\`${command}\` CLI が見つかりません。`,
        );
      }

      return unavailableStatus(
        "codex",
        PROVIDER_NAME,
        "cli",
        `状態確認に失敗しました: ${(error as Error).message}`,
      );
    }
  },

  async summarize(book) {
    const settings = getEffectiveSettings();
    const command = settings.cli.codexCommand || "codex";
    const model = settings.cli.codexModel;
    const status = await this.getStatus();
    if (!status.available || !status.loggedIn) {
      throw new Error(status.detail);
    }

    const tempDir = await mkdtemp(join(tmpdir(), "oreilly-codex-"));
    const schemaPath = join(tempDir, "summary-schema.json");
    const outputPath = join(tempDir, "summary-output.json");

    try {
      await writeFile(schemaPath, JSON.stringify(SUMMARY_SCHEMA, null, 2), "utf8");

      await runCommand(
        command,
        [
          "-a",
          "never",
          "exec",
          ...(model ? ["-m", model] : []),
          "--skip-git-repo-check",
          "--ephemeral",
          "--sandbox",
          "read-only",
          "--output-schema",
          schemaPath,
          "--output-last-message",
          outputPath,
          buildSummaryPrompt(book),
        ],
        {
          timeoutMs: 180_000,
        },
      );

      const output = await readFile(outputPath, "utf8");
      return parseSummaryJson(output.trim());
    } finally {
      await rm(tempDir, { recursive: true, force: true });
    }
  },

  async chatChapter(book, chapter, messages) {
    const settings = getEffectiveSettings();
    const command = settings.cli.codexCommand || "codex";
    const model = settings.cli.codexModel;
    const status = await this.getStatus();
    if (!status.available || !status.loggedIn) {
      throw new Error(status.detail);
    }

    const tempDir = await mkdtemp(join(tmpdir(), "oreilly-codex-chat-"));
    const outputPath = join(tempDir, "chat-output.txt");

    try {
      await runCommand(
        command,
        [
          "-a",
          "never",
          "exec",
          ...(model ? ["-m", model] : []),
          "--skip-git-repo-check",
          "--ephemeral",
          "--sandbox",
          "read-only",
          "--output-last-message",
          outputPath,
          buildChapterChatPrompt(book, chapter, messages),
        ],
        {
          timeoutMs: 180_000,
        },
      );

      const output = await readFile(outputPath, "utf8");
      return output.trim();
    } finally {
      await rm(tempDir, { recursive: true, force: true });
    }
  },
};

function normalizeStatusDetail(text: string): string {
  return text
    .split("\n")
    .map((line) => line.trim())
    .filter((line) => line && !line.startsWith("WARNING:"))
    .join(" ");
}

File: backend/app/services/providers/codex.py
import json
from pathlib import Path
from tempfile import TemporaryDirectory

from app.services.prompts import SUMMARY_SCHEMA, build_chat_prompt, build_summary_prompt
from app.services.providers.base import ProviderStatus, unavailable_status
from app.services.providers.shared import parse_summary_json, run_command
from app.services.settings import get_effective_settings


async def get_status() -> ProviderStatus:
    settings = get_effective_settings()
    command = settings["cli"]["codex_command"] or "codex"
    try:
        stdout, stderr = await run_command(command, "login", "status", timeout=10)
    except FileNotFoundError:
        return unavailable_status("codex", "Codex", "cli", f"`{command}` CLI が見つかりません。")
    except Exception as error:
        return unavailable_status("codex", "Codex", "cli", f"状態確認に失敗しました: {error}")

    text = f"{stdout}\n{stderr}"
    logged_in = "logged in" in text.lower()
    return ProviderStatus(
        id="codex",
        name="Codex",
        kind="cli",
        available=True,
        logged_in=logged_in,
        detail=text.strip() if logged_in else f"未ログイン。`{command} login` が必要です。",
    )


async def summarize(book: dict) -> dict:
    status = await get_status()
    if not status.available or not status.logged_in:
        raise RuntimeError(status.detail)
    settings = get_effective_settings()
    command = settings["cli"]["codex_command"] or "codex"
    model = settings["cli"]["codex_model"]
    with TemporaryDirectory(prefix="mokuji-codex-") as temp_dir:
        schema_path = Path(temp_dir) / "summary-schema.json"
        output_path = Path(temp_dir) / "summary-output.json"
        schema_path.write_text(json.dumps(SUMMARY_SCHEMA), encoding="utf-8")
        args = [
            "-a",
            "never",
            "exec",
            "--skip-git-repo-check",
            "--ephemeral",
            "--sandbox",
            "read-only",
            "--output-schema",
            str(schema_path),
            "--output-last-message",
            str(output_path),
            build_summary_prompt(book),
        ]
        if model:
            args[3:3] = ["-m", model]
        await run_command(command, *args, timeout=180)
        return parse_summary_json(output_path.read_text(encoding="utf-8").strip())


async def chat(book: dict, chapter: dict, messages: list[dict]) -> str:
    status = await get_status()
    if not status.available or not status.logged_in:
        raise RuntimeError(status.detail)
    settings = get_effective_settings()
    command = settings["cli"]["codex_command"] or "codex"
    model = settings["cli"]["codex_model"]
    with TemporaryDirectory(prefix="mokuji-codex-chat-") as temp_dir:
        output_path = Path(temp_dir) / "chat-output.txt"
        args = [
            "-a",
            "never",
            "exec",
            "--skip-git-repo-check",
            "--ephemeral",
            "--sandbox",
            "read-only",
            "--output-last-message",
            str(output_path),
            build_chat_prompt(book, chapter, messages),
        ]
        if model:
            args[3:3] = ["-m", model]
        await run_command(command, *args, timeout=180)
        return output_path.read_text(encoding="utf-8").strip()
